lib/widgets/autocomplete.dart

```diff
@@ -1,43 +1,139 @@
+      return MentionAutocompleteView.init(store: store, narrow: narrow);
+    },
+    itemBuilder: (context, index, option) {
+      Widget avatar;
+      String label;
+      switch (option) {
+        case UserMentionAutocompleteResult(:var userId):
+          avatar = Avatar(userId: userId, size: 32, borderRadius: 3);
+          label = PerAccountStoreWidget.of(context).users[userId]!.fullName;
+        default:
+          avatar = const SizedBox();
+          label = '';
+      }
+      return InkWell(
+        onTap: () {
+          // Probably the same intent that brought up the option that was tapped.
+          // If not, it still shouldn't be off by more than the time it takes
+          // to compute the autocomplete results, which we do asynchronously.
+          final intent = controller.autocompleteIntent();
+          if (intent == null) {
+            return; // Shrug.
+          }
+
+          final store = PerAccountStoreWidget.of(context);
+          final String replacementString;
+          switch (option) {
+            case UserMentionAutocompleteResult(:var userId):
+              // TODO(i18n) language-appropriate space character; check active keyboard?
+              //   (maybe handle centrally in `controller`)
+              replacementString = '${mention(store.users[userId]!, silent: intent.query.silent, users: store.users)} ';
+            default:
+              replacementString = '';
+          }
+
+          controller.value = intent.textEditingValue.replaced(
+            TextRange(
+              start: intent.syntaxStart,
+              end: intent.textEditingValue.selection.end),
+            replacementString,
+          );
+        },
+        child: Padding(
+          padding: const EdgeInsets.symmetric(horizontal: 16.0, vertical: 8.0),
+          child: Row(
+            children: [
+              avatar,
+              const SizedBox(width: 8),
+              Text(label)])));
+    });
+}
+
+class TopicAutocomplete extends AutocompleteField<TopicAutocompleteQuery, TopicAutocompleteResult, Topic> {
+  TopicAutocomplete({
+    super.key,
+    required int streamId,
+    required ComposeTopicController controller,
+    required super.focusNode,
+    required FocusNode contentFocusNode,
+    required super.fieldViewBuilder,
+  }) : super(
+    controller: controller,
+    getAutocompleteIntent: () => controller.autocompleteIntent(),
+    viewModelBuilder: (context) {
+      final store = PerAccountStoreWidget.of(context);
+      return TopicAutocompleteView.init(store: store, streamId: streamId);
+    },
+    itemBuilder: (context, index, option) => InkWell(
+      onTap: () {
+        final intent = controller.autocompleteIntent();
+        if (intent == null) return;
+        final label = option.topic.value;
+        controller.value = intent.textEditingValue.replaced(TextRange(
+          start: intent.syntaxStart,
+          end: intent.textEditingValue.text.length), label);
```

Comment:
Bump on getting test coverage for this behavior: https://github.com/zulip/zulip-flutter/pull/627#discussion_r1685049138